File: 22707281_NguyenXuanChuc_Lab03/22707281_NguyenXuanChuc_Lab03_Layer/backend/src/main/java/iuh/fit/backend/points/service/PointService.java
package iuh.fit.backend.points.service;

import iuh.fit.backend.points.dto.LeaderboardItemResponse;
import iuh.fit.backend.points.dto.PointBalanceResponse;
import iuh.fit.backend.points.dto.PointRequest;
import iuh.fit.backend.points.dto.PointTransactionResponse;

import java.util.List;

public interface PointService {
    PointTransactionResponse earn(PointRequest request);

    PointTransactionResponse deduct(PointRequest request);

    PointBalanceResponse getBalance(String studentCode);

    List<PointTransactionResponse> getHistory(String studentCode);

    List<LeaderboardItemResponse> getLeaderboard(int limit);
}

File: 22707281_NguyenXuanChuc_Lab03/22707281_NguyenXuanChuc_Lab03_Layer/frontend/src/services/pointService.ts
import { apiClient, unwrap } from "./apiClient";
import type { LeaderboardItem, PointBalance, PointPayload, PointTransaction } from "../types";

export const pointService = {
    earn: async (payload: PointPayload) =>
        unwrap<PointTransaction>((await apiClient.post("/api/points/earn", payload))),
    deduct: async (payload: PointPayload) =>
        unwrap<PointTransaction>((await apiClient.post("/api/points/deduct", payload))),
    getBalance: async (studentCode: string) =>
        unwrap<PointBalance>((await apiClient.get(`/api/points/balance/${studentCode}`))),
    getHistory: async (studentCode: string) =>
        unwrap<PointTransaction[]>((await apiClient.get(`/api/points/history/${studentCode}`))),
    getLeaderboard: async (limit = 10) =>
        unwrap<LeaderboardItem[]>((await apiClient.get("/api/points/leaderboard", { params: { limit } }))),
};
